src/messaging/ExchangeMessageDispatch.cpp

```diff
@@ -51,45 +51,60 @@ CHIP_ERROR ExchangeMessageDispatch::SendMessage(SessionManager * sessionManager,
-
-        ReturnErrorOnFailure(sessionManager->PrepareMessage(session, payloadHeader, std::move(message), entryOwner->retainedBuf));
-        CHIP_ERROR err = sessionManager->SendPreparedMessage(session, entryOwner->retainedBuf);
-        err            = ReliableMessageMgr::MapSendError(err, exchangeId, isInitiator);
-        ReturnErrorOnFailure(err);
-        reliableMessageMgr->StartRetransmision(entryOwner.release());
+        // If there is a pending acknowledgment piggyback it on this message.
+        if (reliableMessageContext->HasPiggybackAckPending())
+        {
+            payloadHeader.SetAckMessageCounter(reliableMessageContext->TakePendingPeerAckMessageCounter());
+        }
+
+        if (IsReliableTransmissionAllowed() && reliableMessageContext->AutoRequestAck() &&
+            reliableMessageContext->GetReliableMessageMgr() != nullptr && isReliableTransmission)
+        {
+            auto * reliableMessageMgr = reliableMessageContext->GetReliableMessageMgr();
+
+            payloadHeader.SetNeedsAck(true);
+
+            ReliableMessageMgr::RetransTableEntry * entry = nullptr;
+
+            // Add to Table for subsequent sending
+            ReturnErrorOnFailure(reliableMessageMgr->AddToRetransTable(reliableMessageContext, &entry));
+            auto deleter = [reliableMessageMgr](ReliableMessageMgr::RetransTableEntry * e) {
+                reliableMessageMgr->ClearRetransTable(*e);
+            };
+            std::unique_ptr<ReliableMessageMgr::RetransTableEntry, decltype(deleter)> entryOwner(entry, deleter);
+
+            ReturnErrorOnFailure(
+                sessionManager->PrepareMessage(session, payloadHeader, std::move(message), entryOwner->retainedBuf));
+            CHIP_ERROR err = sessionManager->SendPreparedMessage(session, entryOwner->retainedBuf);
+            err            = ReliableMessageMgr::MapSendError(err, exchangeId, isInitiator);
+            ReturnErrorOnFailure(err);
+            reliableMessageMgr->StartRetransmision(entryOwner.release());
+        }
+        else
+        {
+            ReturnErrorOnFailure(PrepareAndSendMessage(sessionManager, session, payloadHeader, std::move(message)));
+        }
     }
     else
     {
-        // If the channel itself is providing reliability, let's not request MRP acks
-        payloadHeader.SetNeedsAck(false);
-        EncryptedPacketBufferHandle preparedMessage;
-        ReturnErrorOnFailure(sessionManager->PrepareMessage(session, payloadHeader, std::move(message), preparedMessage));
-        ReturnErrorOnFailure(sessionManager->SendPreparedMessage(session, preparedMessage));
+        ReturnErrorOnFailure(PrepareAndSendMessage(sessionManager, session, payloadHeader, std::move(message)));
     }
 
     return CHIP_NO_ERROR;
 }
 
+CHIP_ERROR ExchangeMessageDispatch::PrepareAndSendMessage(SessionManager * sessionManager, const SessionHandle & session,
```

Comment:
Still not convinced all this surgery in this file is needed, but if we do it we should be clear about what sort of message is being sent here:
```suggestion
CHIP_ERROR ExchangeMessageDispatch::SendNonMRPMessage(SessionManager * sessionManager, const SessionHandle & session,
```

Reply:
Changed it to PrepareAndSendNonMRPMessage